spec/integration/parser/epp_parameters_spec.rb

```diff
@@ -0,0 +1,93 @@
+require 'spec_helper'
+require 'puppet_spec/language'
+
+module PuppetSpec
+module EppRunner
+  extend RSpec::Matchers::DSL
+
+  def self.collect_log(source, args, code)
+    Puppet[:code] = code
+    compiler = Puppet::Parser::Compiler.new(Puppet::Node.new('specification'))
+    logs = []
+    Puppet::Util::Log.with_destination(Puppet::Test::LogCollector.new(logs)) do
+      compiler.compile
+      Puppet::Pops::Evaluator::EppEvaluator.inline_epp(compiler.topscope, source, args)
+    end
+    logs = logs.select { |log| log.level == :notice }.map { |log| log.message }
+    logs
+  end
+
+  def run_expectations(decl, expectations, code = 'undef')
+    expectations.each do |args, result|
+      if result[0].is_a?(Class)
+        it "'#{decl}' called with #{args} should fail with error #{result}" do
+          expect {EppRunner.collect_log(decl, args, code)}.to raise_error(result[0], result[1])
+        end
+      else
+        it "'#{decl}' called with #{args} should produce #{result}" do
+          expect(EppRunner.collect_log(decl, args, code)).to include(*result)
+        end
+      end
+    end
+  end
+end
+
+describe "EPP parameter default expressions" do
+  extend EppRunner
+
+  context "to the left of a parameter can use that parameter's value" do
+    run_expectations('<%| $a = 10, $b = $a |%><%= notice($a) %><%= notice($b) %>',
+      [ [{}, ['10', '10']],
+        [{'a' => 2}, ['2', '2']],
+        [{'a' => 2, 'b' => 5}, ['2', '5'] ]])
```

Comment:
This would be greatly enhanced if there was a string in each expectation that explained that example. Here the first would be "$b is assigned default value of $a", the second "$b is assigned the given value of $a", etc. This helps both reading the examples (now it took a while to figure out what the example meant. The method run_expectations should then use the given string in the "it" - to make it easier to find which one of the expectations that failed (since the failure is reported by rspec against the line where the "it" is. (I had the pleasure of fixing > 100 such failures on an API change in another rspec test using something similar.